### DISCUSSION TITLE
Update: MPI-parallel working version

### DISCUSSION THREAD

**Discussion by fdrmrc** [fdrmrc/Polydeal#84] [Show and tell]
The present version now support also distributed memory parallelism. It has been tested both in 2D and 3D on distributed grids with different number of processors and configurations. @andreacangiani 

Here are some first working preliminary results:
- $\Omega=[0,1]^d$,
- $u=\Pi_{i=1}^{d}\sin(2 \pi x_i)$,
- Left: partition across processors. Each color represents a processor which owns that portion of triangulation (12 in these tests),
- Right: interpolation on the fine triangulation of the solution defined on the agglomerated mesh.

$d=2$
![parallel_2D](https://github.com/fdrmrc/Polydeal/assets/50499769/6c93dbd4-dd48-4e62-96d2-c1d84b97d036)

$d=3$
![parallel_3D](https://github.com/fdrmrc/Polydeal/assets/50499769/bc90b55d-df12-4a71-9645-03d47724f08e)

**Comment by andreacangiani** [comment]
Thanks Marco! Seems like great news. Looking forward to more detailed results. Not your fault but the way the mesh is subdivided among processors is weird...

**Reply by fdrmrc** [reply]
Yes the partitions are generated by the library `p4est` which guarantees (among other features) to have load balancing. Then, locally to each partition we are free to agglomerate elements in the way we prefer.

**Comment by fdrmrc** [comment]
@luca-heltai , @andreacangiani

Some more preliminary numbers for the 3D Laplace problem on $[0,1]^3$ (on my laptop):

- Number of processors: 12 
- Fine mesh $\tau_h$: 2.097.152 hexas
- Agglomerated mesh: 180.000 polygons
- Finite Element Space: DGQ of degree $1$ => number of DoFs=180.000*8=1.440.000
- Assembly time: ~60[s]
- Solver: CG preconditioned by AMG
- Time to solution: ~6[s]
- $L^2$ error: 9e-4
